flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -2075,10 +2141,21 @@ public void reportStats(long id, ExecutionAttemptID attemptId, CheckpointMetrics
 
     // ------------------------------------------------------------------------
 
-    private final class ScheduledTrigger implements Runnable {
-
+    final class ScheduledTrigger implements Runnable {
         @Override
         public void run() {
+            synchronized (lock) {
+                long checkpointInterval = getCurrentCheckpointInterval();
+                if (checkpointInterval != Long.MAX_VALUE) {
+                    nextCheckpointTriggeringRelativeTime += checkpointInterval;
```

Comment:
With the current implementation, the difference between the actual time of the next invocation of `ScheduledTrigger()` and the value of `nextCheckpointTriggeringRelativeTime` will likely increase with the number of times this method is executed.

It should be better to use `nextCheckpointTriggeringRelativeTime = clock.relativeTimeMillis() + tillNextMillis` here to avoid accumulation of this difference.